adder/sharding/shard.go

```diff
@@ -0,0 +1,148 @@
+package sharding
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/ipfs-cluster/api"
+
+	humanize "github.com/dustin/go-humanize"
+	rpc "github.com/hsanjuan/go-libp2p-gorpc"
+	cid "github.com/ipfs/go-cid"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+// a shard represents a set of blocks (or bucket) which have been assigned
+// a peer to be block-put and will be part of the same shard in the
+// cluster DAG.
+type shard struct {
```

Comment:
@hsanjuan Why haven't we implemented the [`ipld.Node`](https://github.com/ipfs/go-ipld-format/blob/c74f9ff3b2fce8d81349de8245ce07413b26f64e/format.go#L26) and [`ipld.Resolver`](https://github.com/ipfs/go-ipld-format/blob/c74f9ff3b2fce8d81349de8245ce07413b26f64e/format.go#L12) interfaces for Shard nodes?

Reply:
We use ipld-cbor below, and that node implements those. A shard node is an ipld-cbor node.

Reply:
But why not a ipld-shard node is more the question?

Reply:
because ipld-cbor covers us so why not re-using something that is there to use?

Reply:
that's a good convincing point. It opens some questions about how we are going to serialize things since we will have the choice (protobuf, cbor?). It will require that we register a block type for cluster in https://github.com/multiformats/multicodec

Reply:
I think as a rough idea of what is required the [`ipld-git`](https://github.com/ipfs/go-ipfs-git) format gives us a good idea, though it doesn't answer the question of how to serialize the data and I am not entirely sure about the benefits of cbor compared to using protobuf. I am working through the recovery scenario just to make sure that this work will actually allow us to recover the equivalent of the cluster root dag from just the ipld-shard nodes.